### PR TITLE
[Part1][Live-Scale] Support hot-reload appsetting configs

### PR DESCRIPTION
**Pull request by JialinXin**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR/ServiceOptions.cs [hunk 1]
```diff
@@ -76,5 +76,15 @@ public class ServiceOptions : IServiceEndpointOptions
         /// Gets or sets the proxy used when ServiceEndpoint will attempt to connect to Azure SignalR.
         /// </summary>
         public IWebProxy Proxy { get; set; }
+
+        /// <summary>
+        /// Gets or sets the the flag to enable live-scale Azure SignalR Endpoints with updating configuration.
+        /// </summary>
+        public bool EnableAutoScale { get; set; } = false;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceOptions.cs]: >EnableAutoScale [](start = 20, length = 15)

naming: `EnableAzureSignalREndpointAutoScale`?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: extra line

File: test/Microsoft.Azure.SignalR.Tests/AddAzureSignalRFacts.cs [hunk 3]
```diff
@@ -369,5 +371,41 @@ public void AddAzureSignalRCustomizeEndpointsOverridesConfigValue(string default
...
+                Assert.Equal(2, options.Endpoints.Length);
+                Assert.True(options.EnableAutoScale);
+
+                // Update config file to add a new endpoint ConnectionString
+                var text = File.ReadAllText(ConfigFile);
+                text = text.Replace("\"ConnectionString\": [", "\"ConnectionString\": [ {\"EP3:Primary\": \"Endpoint = https://customconnectionstring;AccessKey=1\"},");
+                File.WriteAllText(ConfigFile, text);
+
+                // give a few seconds delay for change detected
+                await Task.Delay(2000);
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests/AddAzureSignalRFacts.cs]: >2000 [](start = 33, length = 4)

is it too much?

File: test/Microsoft.Azure.SignalR.Tests/AddAzureSignalRFacts.cs [hunk 4]
```diff
@@ -369,5 +371,41 @@ public void AddAzureSignalRCustomizeEndpointsOverridesConfigValue(string default
...
+
+                var optionsMonitor = serviceProvider.GetService<IOptionsMonitor<ServiceOptions>>();
+                var options = optionsMonitor.CurrentValue;
+
+                Assert.Equal(2, options.Endpoints.Length);
+                Assert.True(options.EnableAutoScale);
+
+                // Update config file to add a new endpoint ConnectionString
+                var text = File.ReadAllText(ConfigFile);
+                text = text.Replace("\"ConnectionString\": [", "\"ConnectionString\": [ {\"EP3:Primary\": \"Endpoint = https://customconnectionstring;AccessKey=1\"},");
```
Comment (vicancy) [inline test/Microsoft.Azure.SignalR.Tests/AddAzureSignalRFacts.cs]: >Replace [](start = 28, length = 7)

This sounds error prone, use configuration write instead?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        private readonly ILogger _logger;
+        private bool _inited = false;
+
+        private IReadOnlyList<ServiceEndpoint> _endpointsStore = new List<ServiceEndpoint>();
+
+        public ServiceScaleManager(ILoggerFactory loggerFactory,
+            IOptionsMonitor<ServiceOptions> optionsMonitor)
+        {
+            _logger = loggerFactory?.CreateLogger<ServiceScaleManager>();
+            OnChange(optionsMonitor.CurrentValue);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: >OnChange [](start = 12, length = 8)

no need to call as `_inited` is `false` here?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        private Task OnRemove(IReadOnlyList<ServiceEndpoint> endpoints)
+        {
+            throw new NotImplementedException();
+        }
+
+        private (IReadOnlyList<ServiceEndpoint> AddedEndpoints, IReadOnlyList<ServiceEndpoint> RemovedEndpoints)
+            GetChangedEndpoints(IReadOnlyList<ServiceEndpoint> updatedEndpoints)
+        {
+            // Compare by ConnectionString
+            var cachedIds = _endpointsStore.Select(e => e.ConnectionString).ToList();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: >Select [](start = 44, length = 6)

no need to select, `ServiceEndpoint` overrides `Equals` already

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                Log.DetectEndpointChanges(_logger);
+
+                var updatedEndpoints = options.Endpoints;
+                var endpoints = GetChangedEndpoints(updatedEndpoints);
+
+                // Add then remove to minor affect to new clients
+                OnAdd(endpoints.AddedEndpoints);
+
+                OnRemove(endpoints.RemovedEndpoints);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: >OnRemove [](start = 16, length = 8)

`add` and `remove` can be done parallelly?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class ServiceScaleManager
+    {
+        private readonly ILogger _logger;
+        private bool _inited = false;
+
+        private IReadOnlyList<ServiceEndpoint> _endpointsStore = new List<ServiceEndpoint>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: >_endpointsStore [](start = 47, length = 15)

`private volatile IReadOnlyList<ServiceEndpoint> _endpointsStore = null;`?

Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: >new List<ServiceEndpoint>() [](start = 65, length = 27)

`null` is enough?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs (hunk 7 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: I think better to wait add complete cause add need to wait server match before adding negotiation and remove will remove negotiation directly, which means before adding successfully, there'll be few endpoints available.

---
In reply to: [368272199](https://github.com/Azure/azure-signalr/pull/801#discussion_r368272199) [](ancestors = 368272199)

File: src/Microsoft.Azure.SignalR/DependencyInjectionExtensions.cs [hunk 9]
```diff
@@ -22,15 +23,20 @@ public static class AzureSignalRDependencyInjectionExtensions
...
         /// </remarks>
-        public static ISignalRServerBuilder AddAzureSignalR(this ISignalRServerBuilder builder)
+        public static ISignalRServerBuilder AddAzureSignalR(this ISignalRServerBuilder builder, IConfiguration configuration = null)
         {
-
             builder.Services.AddSingleton<IConfigureOptions<ServiceOptions>, ServiceOptionsSetup>();
+            if (configuration != null)
+            {
+                // supports hot-reload settings.
+                builder.Services.AddSingleton<IOptionsChangeTokenSource<ServiceOptions>>(new ConfigurationChangeTokenSource<ServiceOptions>(configuration));
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/DependencyInjectionExtensions.cs]: how about `ServiceOptionsSetup` to implement `IOptionsChangeTokenSource`?

File: src/Microsoft.Azure.SignalR/ServiceOptions.cs [hunk 10]
```diff
@@ -76,5 +76,10 @@ public class ServiceOptions : IServiceEndpointOptions
         /// Gets or sets the proxy used when ServiceEndpoint will attempt to connect to Azure SignalR.
         /// </summary>
         public IWebProxy Proxy { get; set; }
+
+        /// <summary>
+        /// Specifies the timeout of scaling endpoints in running app (in minutes). 
+        /// </summary>
+        public TimeSpan ServiceScaleTimeout { get; set; } = TimeSpan.FromMinutes(Constants.DefaultScaleTimeoutInMinutes);
```
Comment (terencefan) [inline src/Microsoft.Azure.SignalR/ServiceOptions.cs]: Maybe we should leave this option internal since the feature the PR is only the very first part of the feature?

File: build/dependencies.props [hunk 11]
```diff
@@ -21,7 +21,7 @@
     <MicrosoftExtensionsLoggingPackageVersion>2.1.0</MicrosoftExtensionsLoggingPackageVersion>
     <MicrosoftExtensionsPrimitivesPackageVersion>2.1.0</MicrosoftExtensionsPrimitivesPackageVersion>
     <MicrosoftAspNetCoreSignalRClient>1.1.0</MicrosoftAspNetCoreSignalRClient>
-    
+
```
Comment (vicancy) [inline build/dependencies.props]: revert?

File: samples/ChatSample/ChatSample/appsettings.json [hunk 12]
```diff
@@ -1,7 +1,14 @@
 ﻿{
   "Azure": {
     "SignalR": {
-      "ConnectionString": ""
+      "ConnectionString": [
+        {
+          "EP1:Primary": ""
```
Comment (vicancy) [inline samples/ChatSample/ChatSample/appsettings.json]: revert?

File: src/Microsoft.Azure.SignalR.Common/Constants.cs [hunk 13]
```diff
@@ -10,8 +10,10 @@ internal static class Constants
         public const string ServerStickyModeDefaultKey = "Azure:SignalR:ServerStickyMode";
         public const string ConnectionStringDefaultKey = "Azure:SignalR:ConnectionString";
         public const string ApplicationNameDefaultKey = "Azure:SignalR:ApplicationName";
+        public const string EnableAutoScaleDefaultKey = "Azure:SignalR:EnableAutoScale";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Constants.cs]: `"Azure:SignalR:AutoScale": true`?

File: src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs [hunk 14]
```diff
@@ -38,12 +39,14 @@ internal class ServiceHubDispatcher<THub> where THub : Hub
...
             IServerNameProvider nameProvider,
-            IClientConnectionFactory clientConnectionFactory)
+            IClientConnectionFactory clientConnectionFactory,
+            ServiceScaleManager serviceScaleManager)
         {
             _serviceProtocol = serviceProtocol;
             _serviceConnectionManager = serviceConnectionManager;
             _clientConnectionManager = clientConnectionManager;
             _serviceEndpointManager = serviceEndpointManager;
+            _serviceScaleManager = serviceScaleManager;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs]: if `ServiceScaleManager ` is a dependency, use interface? 
https://stackoverflow.com/questions/43079277/do-we-need-interfaces-for-dependency-injection
> Making your consumer classes depend on interfaces (abstractions) instead of other concrete classes makes them much easier to test.

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            _logger = loggerFactory?.CreateLogger<ServiceScaleManager>();
+
+            // Disable optionsMonitor until feature ready.
+            // optionsMonitor.OnChange(OnChange);
+
+            _endpointsStore = optionsMonitor.CurrentValue.Endpoints;
+        }
+
+        private void OnChange(ServiceOptions options)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: If the configuration is changed intensively during some period... shall the new detected Change cancel the previous waiting and begin the latest sync?

File: samples/ChatSample/ChatSample/appsettings.json (hunk 12 above)
Comment (JialinXin) [reply samples/ChatSample/ChatSample/appsettings.json]: sample to show how to configure multiple endpoints through ConnectionString.

---
In reply to: [374090429](https://github.com/Azure/azure-signalr/pull/801#discussion_r374090429) [](ancestors = 374090429)

File: src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs (hunk 14 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs]: Sure. The interface is added in later PRs.

---
In reply to: [374094131](https://github.com/Azure/azure-signalr/pull/801#discussion_r374094131) [](ancestors = 374094131)

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs (hunk 15 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/EndpointProvider/ServiceScaleManager.cs]: For this async endpoints change, it's not good to cancel previous operation and hard to rollback some changes. This may lead to some available endpoints inconsistent in `ServiceEndpointManager` and `MultiEndpointServiceConnectionContainer` with some unexpected behaviors.
* For add, probably only final add negotiation is waiting and can be cancelled, but some server connections are already established.
* For remove, offline call is triggered to drop clients and remove server routing, probably only server connection stop and clean is able to cancel. 
So I think it's good to respect every time configuration change. App server should be aware of the impact of value update.

File: src/Microsoft.Azure.SignalR.Common/Constants.cs (hunk 13 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR.Common/Constants.cs]: Enable by default and reverted the change.

File: build/dependencies.props (hunk 11 above)
Comment (JialinXin) [reply build/dependencies.props]: remove the spaces in empty lines.

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 16]
```diff
@@ -30,5 +44,74 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+            _endpointsStore = updatedEndpoints.ToList();
+        }
+
+        private Task DoScaleAsync(IReadOnlyList<ServiceEndpoint> addedEndpoints, IReadOnlyList<ServiceEndpoint> removedEndpoints)
+        {
+            // First add then remove to minor the affect to new clients.
+            // Update EndpointType follow same process.
+            // AddServiceEndpoints(addedEndpoints);
+            // 
+            // RemoveServiceEndpoint(removedEndpoints);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: >RemoveServiceEndpoint [](start = 15, length = 21)

if it is not enabled, shall we move it and its implementation to another PR with tests ready?

File: src/Microsoft.Azure.SignalR/ServiceOptions.cs [hunk 17]
```diff
@@ -76,5 +76,11 @@ public class ServiceOptions : IServiceEndpointOptions
         /// Gets or sets the proxy used when ServiceEndpoint will attempt to connect to Azure SignalR.
         /// </summary>
         public IWebProxy Proxy { get; set; }
+
+        /// <summary>
+        /// Specifies the timeout of scaling endpoints in running app (in seconds). 
+        /// </summary>
+        // TODO: make public when feature ready.
+        internal TimeSpan ServiceScaleTimeout { get; set; } = TimeSpan.FromSeconds(Constants.DefaultScaleTimeoutInSeconds);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceOptions.cs]: >ServiceScaleTimeout [](start = 26, length = 19)

remove if this PR does not use it

**Comment by vicancy** [comment]
>   <PropertyGroup Label="Package Versions">

no change, revert

---
Refers to: build/dependencies.props:5 in ef0b095. [](commit_id = ef0b095d5b8fbc41ec944ae70603d3a4e525d6cb, deletion_comment = False)

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 18]
```diff
@@ -2,23 +2,37 @@
...
+        public ServiceEndpointManager(IOptionsMonitor<ServiceOptions> optionsMonitor, ILoggerFactory loggerFactory) :
+            base(optionsMonitor.CurrentValue, loggerFactory.CreateLogger<ServiceEndpointManager>())
         {
             if (Endpoints.Length == 0)
             {
                 throw new ArgumentException(ServiceEndpointProvider.ConnectionStringNotFound);
             }
-            _options = options;
+            _options = optionsMonitor.CurrentValue;
+            _logger = loggerFactory?.CreateLogger<ServiceEndpointManager>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: >?. [](start = 35, length = 2)

`?? NullLogger...`

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 19]
```diff
@@ -30,5 +44,74 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+
+        private void OnChange(ServiceOptions options)
+        {
+            Log.DetectConfigurationChanges(_logger);
+
+            // Reset local cache and validate result
+            SetValuableEndpoints(GetEndpoints(options));
+            if (Endpoints.Length == 0)
+            {
+                throw new ArgumentException(ServiceEndpointProvider.ConnectionStringNotFound);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: >ArgumentException [](start = 26, length = 17)

what's the behavior when OnChange throw?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 20]
```diff
@@ -30,5 +44,74 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+            }
+            var updatedEndpoints = Endpoints.ToList();
+
+            var endpoints = GetChangedEndpoints(updatedEndpoints);
+
+            // RenameServiceEndpoints(endpoints.RenamedEndpoints);
+
+            _ = DoScaleAsync(endpoints.AddedEndpoints, endpoints.RemovedEndpoints);
+
+            _endpointsStore = updatedEndpoints.ToList();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: >updatedEndpoints [](start = 30, length = 16)

it is `ToList` twice?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 21]
```diff
@@ -30,5 +44,74 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+            var renamedEndpoints = updatedEndpoints.Except(_endpointsStore).Except(addedEndpoints).ToList();
+
+            return (AddedEndpoints: addedEndpoints, RemovedEndpoints: removedEndpoints, RenamedEndpoints: renamedEndpoints);
+        }
+
+        private sealed class ServiceEndpointWeakComparer : IEqualityComparer<ServiceEndpoint>
+        {
+            public bool Equals(ServiceEndpoint x, ServiceEndpoint y)
+            {
+                return x.ConnectionString == y.ConnectionString && x.EndpointType == y.EndpointType;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: >ConnectionString [](start = 25, length = 16)

ConnectionString contains the access key? are the 2 ServiceEndpoints considered as the same one if one is using primary access key and another is using secondary access key?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs [hunk 22]
```diff
@@ -12,11 +12,13 @@ namespace Microsoft.Azure.SignalR
 {
     internal abstract class ServiceEndpointManagerBase : IServiceEndpointManager
     {
+        // Endpoints for negotiation
         private readonly ConcurrentDictionary<string, IReadOnlyList<HubServiceEndpoint>> _endpointsPerHub = new ConcurrentDictionary<string, IReadOnlyList<HubServiceEndpoint>>();
 
         private readonly ILogger _logger;
 
-        public ServiceEndpoint[] Endpoints { get; }
+        // Filtered valuable endpoints from ServiceOptions
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs]: >valuable [](start = 20, length = 8)

what does `valuable` mean?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs (hunk 21 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: Right. Previously thought about regenerated keys and just realized no need to reconnect even in this case, so Endpoint compare should be enough. Will update.

---
In reply to: [377514199](https://github.com/Azure/azure-signalr/pull/801#discussion_r377514199) [](ancestors = 377514199)

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs (hunk 22 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs]: Follow original comments about this. Suppose means filter out duplicate endpoints from all available configuration sources.
> // select the most valuable endpoint with the same endpoint address


---
In reply to: [377514575](https://github.com/Azure/azure-signalr/pull/801#discussion_r377514575) [](ancestors = 377514575)

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs (hunk 19 above)
Comment (JialinXin) [reply src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: Good question. Skip changes and log error probably friendly? 

---
In reply to: [377512292](https://github.com/Azure/azure-signalr/pull/801#discussion_r377512292) [](ancestors = 377512292)

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 23]
```diff
@@ -2,23 +2,37 @@
...
             {
                 throw new ArgumentException(ServiceEndpointProvider.ConnectionStringNotFound);
             }
-            _options = options;
+            _options = optionsMonitor.CurrentValue;
+            _logger = loggerFactory?.CreateLogger<ServiceEndpointManager>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+
+            // TODO: Enable optionsMonitor.OnChange when feature ready.
+            // optionsMonitor.OnChange(OnChange);
+            _endpointsStore = Endpoints.ToList();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: no need to `ToList` as `Endpoints` is immutable?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 24]
```diff
@@ -30,5 +44,68 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+            if (endpoints.Length == 0)
+            {
+                Log.EndpointNotFound(_logger);
+                return; 
+            }
+            Endpoints = endpoints;
+
+            var updatedEndpoints = GetChangedEndpoints(Endpoints);
+
+            _endpointsStore = Endpoints.ToList();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: is reusing `Endpoints` enough?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 25]
```diff
@@ -30,5 +44,68 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+        }
+
+        private (IReadOnlyList<ServiceEndpoint> AddedEndpoints, 
+            IReadOnlyList<ServiceEndpoint> RemovedEndpoints,
+            IReadOnlyList<ServiceEndpoint> RenamedEndpoints)
+            GetChangedEndpoints(IEnumerable<ServiceEndpoint> updatedEndpoints)
+        {
+            var addedEndpoints = updatedEndpoints.Except(_endpointsStore, new ServiceEndpointWeakComparer()).ToList();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: `var original = _endpointsStore` and use this local var to compare?

File: src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs [hunk 26]
```diff
@@ -30,5 +44,68 @@ public override IServiceEndpointProvider GetEndpointProvider(ServiceEndpoint end
...
+        private sealed class ServiceEndpointWeakComparer : IEqualityComparer<ServiceEndpoint>
+        {
+            public bool Equals(ServiceEndpoint x, ServiceEndpoint y)
+            {
+                return x.Endpoint == y.Endpoint && x.EndpointType == y.EndpointType;
+            }
+
+            public int GetHashCode(ServiceEndpoint obj)
+            {
+                return obj.ConnectionString.GetHashCode() ^ obj.EndpointType.GetHashCode();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/EndpointProvider/ServiceEndpointManager.cs]: `ConnectionString` -> `Endpoint`?
